File: leetcode/875. Koko Eating Bananas/main.go
package main

import "cmp"

func maxInArray[T cmp.Ordered](arr []T) T {
	maxNum := arr[0]

	for i := 1; i < len(arr); i++ {
		if maxNum < arr[i] {
			maxNum = arr[i]
		}
	}

	return maxNum
}

func minEatingSpeed(piles []int, h int) int {
	var canEat func(int) bool
	canEat = func(k int) bool {
		actH := 0

		for _, pile := range piles {
			actH += (pile + k - 1) / k
		}

		return actH <= h
	}

	left, right := 1, maxInArray(piles)

	for left < right {
		mid := (left + right) / 2
		if canEat(mid) {
			right = mid
		} else {
			left = mid + 1
		}
	}

	return left
}

File: leetcode/875. Koko Eating Bananas/main.rs
impl Solution {
    fn can_eat(k: i32, piles: &[i32], h: i32) -> bool {
        let mut act_h = 0;

        for &pile in piles {
            act_h += (pile + k - 1) / k
        }

        act_h <= h
    }

    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        let mut left = 1;
        let mut right = *piles.iter().max().unwrap();

        while left < right {
            let mid = left + (right - left) / 2;
            if Solution::can_eat(mid, &piles, h) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        left
    }
}
